Introduction to Regular Expressions

jamos-tay opened the pull request.

jamos-tay committed:
Add skeleton

jamos-tay committed:
Fixed messed up links

jamos-tay committed:
Missing <br>s

jamos-tay committed:
Update 1

jamos-tay committed:
Update 1 fix formatting

jamos-tay committed:
More fixes

jamos-tay committed:
Add todos

jamos-tay committed:
Update

jamos-tay committed:
Update

wkurniawan07 commented on contents/ToC.md:
```diff
@@ -27,4 +27,5 @@
-* AI
\ No newline at end of file
+* AI
+* [Regex](regex/Regex.md)
```
Make the visible text "Regular Expressions" to cater to those who may not know what "Regex" means.

wkurniawan07 commented on contents/regex/Regex.md:
```diff
@@ -0,0 +1,316 @@
+
+## Why should I learn Regex?
+
+* **Regex is everywhere** - Regular Expressions are technically a language by themselves, but they are usually implemented within other languages. Almost any commonly used language, like Java, C++ or Python supports Regex is some fashion (A full list is available [here](http://www.regular-expressions.info/tools.html)). Even most decent IDEs or text editors support Regex in 'Find and replace' oper ...
```
`Almost any` -> `Almost all`.
`Even most decent IDEs or text editors` -> `In addition, many IDEs and text editors`.

wkurniawan07 commented on contents/regex/Regex.md:
```diff
@@ -0,0 +1,316 @@
+* **Regex is everywhere** - Regular Expressions are technically a language by themselves, but they are usually implemented within other languages. Almost any commonly used language, like Java, C++ or Python supports Regex is some fashion (A full list is available [here](http://www.regular-expressions.info/tools.html)). Even most decent IDEs or text editors support Regex in 'Find and replace' oper ...
+* **Regex is portable** - People often complain that switching to a new language usually means having to learn a new API and syntax. Regex however is implemented almost the same way on every single language, so you can just learn it once and use it everywhere.
+* **Regex is useful** - Although Regex doesn't have a specific use, it is a general use library with many different applications. It can be used automate many different common programming tasks such as validating user input, parsing data from text, and string manipulation, all of which are commonly done in any programming project.
+* **Regex is convenient** - Having struggled with many different string handling libraries over the years, I know that string processing is usually a mundane and tedious task. Regex offers a simple and painless way of dealing with them.
```
- ~~Having struggled with many different string handling libraries over the years, I know that~~
- "Regex offers a simple and painless way of dealing **some** with them" because not all string processing are regex-related.

wkurniawan07 commented on contents/regex/Regex.md:
Char**a**cter classes

wkurniawan07 commented on contents/regex/Regex.md:
Missing the closing backtick.

wkurniawan07 commented on contents/regex/Regex.md:
Use list instead of `<br>` separator:
- Item 1
- Item 2
- Item 3

wkurniawan07 commented on contents/regex/Regex.md:
Is this a placeholder?

wkurniawan07 commented on contents/regex/Regex.md:
Use `1.` for all numbers; GFM automatically translates it to proper list numbering.

wkurniawan07 commented on contents/regex/Regex.md:
This falls under greedy operators, right? If it is the case, should make the heading one level lower, i.e. `### Effect on performance`.

wkurniawan07 commented on contents/regex/Regex.md:
Similarly here.

wkurniawan07 commented on contents/regex/Regex.md:
You may want to make mentions to lazy quantifiers here. It's often the solution to the catastrophic backtracking.

jamos-tay replied:
Done =)

wkurniawan07 commented on contents/regex/Regex.md:
Above two sections are not used?

wkurniawan07 commented on contents/regex/Regex.md:
For all above three sections: make it a list instead of `<br>`-delimited.

wkurniawan07 reviewed (changes requested):
Good article on introductions to regex. Sufficient examples provided, but may be good to add some actual programming examples, e.g. with common languages such as Java and JS.

Also, here's an excellent resource to be listed (in fact the one I used to get most of my regex knowledge): http://www.rexegg.com.

damithc commented:
@jamos-tay let me know after approved by peer reviewer

Kenneth-LJS commented on contents/regex/Regex.md:
A good place to start will be to go through how regex works on a basic level before going on to all the features.
E.g. "A regex is a text pattern. It can be used to check if a piece of text follows a specified format (to ensure a string is an email address, etc), to look for the existence of a substring, or to replace substrings.

jamos-tay replied:
Good idea, added a short introduction

Kenneth-LJS commented on contents/regex/Regex.md:
You explained the groups (1, 2, 3), but didn't explain how to use them in the replacing regex (\1, \2 and \3). The usage of / in the date may also confuse the reader with the \ and /s.

Kenneth-LJS commented on contents/regex/Regex.md:
"And not only that..."

Kenneth-LJS reviewed (changes requested):
Overall comprehensive introduction. However, readers new to the concept of regex will find themselves thrown into its complex functionality before they even understand how they are supposed to use it on a basic level. Can be improved to become more newbie-friendly.

Kenneth-LJS commented on contents/regex/Regex.md:
Your table above explains "{number, number}" but not "{number}".

Kenneth-LJS commented on contents/regex/Regex.md:
So, how exactly does Regex work?

Kenneth-LJS commented on contents/regex/Regex.md:
"If it does, check the second character of both strings and so on."

jamos-tay committed:
Made fixes

jamos-tay committed:
Fixed more errors

jamos-tay committed:
Add section

jamos-tay commented:
@wkurniawan07 Thanks for your review 👍 Good link, I added it to the appendix. I think code examples are a good idea, but my intention for this book chapter was to teach Regex without using code, since users may not be familiar with Java or Javascript, so I left them out. I added a 'Further Reading' section, though.

@Kenneth-LJS Thanks for the feedback, I added a simpler search example to start with.

wkurniawan07 commented on contents/regex/Regex.md:
... a~~n~~ string of 8 digits.

wkurniawan07 commented on contents/regex/Regex.md:
Char**a**cter classes

wkurniawan07 commented on contents/regex/Regex.md:
"Used for extracting data, see below" needs not be surrounded by backtick.
Also, what is "below"? I think you're trying to say "above" where you demonstrated the usage of capturing group.

wkurniawan07 commented on contents/regex/Regex.md:
...instead of `.+` and `.*`?

wkurniawan07 commented on contents/regex/Regex.md:
Similarly, ...instead of `.+` and `.*`?

wkurniawan07 commented on contents/regex/Regex.md:
Remove this?

wkurniawan07 reviewed (changes requested):
Sorry for the late review! Minor (mostly cosmetic) changes requested, the content is good otherwise!

jamos-tay committed:
Text fixes

jamos-tay commented:
@wkurniawan07 It's okay, thanks for the review! Have made the necessary changes

wkurniawan07 reviewed (approved):
LGTM 👍

damithc commented on contents/regex/Regex.md:
```diff
@@ -0,0 +1,367 @@
+
+## What are Regular Expressions?
+
+A Regular Expression (or Regex for short) is a string of characters that define a search pattern. You can think of a regular expression as a list of descriptions that describe a string. These strings can be tested against other text formatted data to find, match and extract data.
```
comma after `match`

damithc commented on contents/regex/Regex.md:
```diff
@@ -0,0 +1,367 @@
+
+A Regular Expression (or Regex for short) is a string of characters that define a search pattern. You can think of a regular expression as a list of descriptions that describe a string. These strings can be tested against other text formatted data to find, match and extract data.
+
+Imagine you're a witness of a robbery. How might you describe the robber to the police? You might say that the robber has a beard, brown hair and a facial scar. With this information, the police can round up everyone with beards, brown hair and scars as possible suspects. This is similar to how Regex works. You specify a search pattern using a string (e.g. It starts with 'C', ends with a `S`, and ...
```
Wrong type of quotes around `C`

damithc commented on contents/regex/Regex.md:
```diff
@@ -0,0 +1,367 @@
+
+A Regular Expression (or Regex for short) is a string of characters that define a search pattern. You can think of a regular expression as a list of descriptions that describe a string. These strings can be tested against other text formatted data to find, match and extract data.
+
+Imagine you're a witness of a robbery. How might you describe the robber to the police? You might say that the robber has a beard, brown hair and a facial scar. With this information, the police can round up everyone with beards, brown hair and scars as possible suspects. This is similar to how Regex works. You specify a search pattern using a string (e.g. It starts with 'C', ends with a `S`, and ...
```
The last part of this paragraph sounds like regex is a tool. Is that correct? I thought it is a technique of specifying patterns and the tools can use that technique to implement text searching.

damithc replied:
This is a nice analogy. If you took it from somewhere else, you need to credit the source.

jamos-tay replied:
@damithc Yes, technically Regex refers to the pattern itself (i.e. a regular expression). Have made the change accordingly.

I came up with the analogy myself =)

damithc commented on contents/regex/Regex.md:
```diff
@@ -0,0 +1,367 @@
+
+## Introduction
+
+The power of regex revolves around the idea that most data comes in a regular, predictable format (e.g. serial numbers, collected data, statistics). Some everyday examples are as follows:
```
This sentence gives some examples and then there are more examples in the list below. Just one or the other is enough?

damithc commented on contents/regex/Regex.md:
Use `document` or `article` instead of `guide`? The latter sets up some specific expectations that this document might not meet.

damithc commented on contents/regex/Regex.md:
`have want`

damithc replied:
Use of `length` is a bit confusing. At first I thought you are talking about looking for text of a specific length. Rephrase?

damithc commented on contents/regex/Regex.md:
This the third time we have a list of examples where regex are applicable. :-p

jamos-tay replied:
@damithc For this one I wanted to bring up examples in particular that can't be done with a simple Ctrl-F. Will change it to just going in depth into one example.

damithc commented on contents/regex/Regex.md:
`English`

damithc commented on contents/regex/Regex.md:
`Reges` is a typo?

damithc commented on contents/regex/Regex.md:
This note should be in the next section?

damithc commented on contents/regex/Regex.md:
Do we need to format this as a heading?

damithc commented on contents/regex/Regex.md:
`Many languages implement it differently` : this sounds like a negative thing. Rephrase?

damithc commented on contents/regex/Regex.md:
You are better off pointing to a more comprehensive cheat sheet elsewhere.

damithc commented on contents/regex/Regex.md:
Give an overview of what the reader is going to find in each of these links.

damithc commented on contents/regex/Regex.md:
Give links (at least to the amazon page)?

damithc reviewed (commented):
Good work @jamos-tay 
Some comments added. Revise as you see fit. I did not have time to do a detailed review of the entire doc though. 
Adding more content and/or getting more peer reviews is welcome but not required. 
I will be merging end of week 12

jamos-tay committed:
First review corrections

jamos-tay committed:
Typo

damithc committed:
Merge branch 'master' into master
